src/index/IndexImpl.Text.cpp

```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+      for (const auto& line :
+           getWordsLineFromDocsFile(docsFile, localeManager)) {
+        processLine(line);
+      }
+    }
+  } else {
+    // Case where: !useDocsFileForVocabulary
+    for (const auto& line :
+         WordsFileParser{wordsFile, textVocab_.getLocaleManager()}) {
+      processLine(line);
+    }
+  }
+  lastTextRecordIndexOfNonLiterals_ = currentContext.get();
+
+  // Parse literals if specified
+  if (addWordsFromLiterals) {
+    for (const auto& line : wordsInLiterals(currentContext.get() + 1)) {
+      processLine(line);
+    }
   }
+
+  // Warnings
   if (entityNotFoundErrorMsgCount > 0) {
     LOG(WARN) << "Number of mentions of entities not found in the vocabulary: "
               << entityNotFoundErrorMsgCount << std::endl;
   }
   LOG(DEBUG) << "Number of total entity mentions: " << nofEntityPostings
              << std::endl;
+
+  // Add last entries
   ++nofContexts;
   addContextToVector(vec, currentContext, wordsInContext, entitiesInContext);
+
+  // Save metadata
   textMeta_.setNofTextRecords(nofContexts);
   textMeta_.setNofWordPostings(nofWordPostings);
   textMeta_.setNofEntityPostings(nofEntityPostings);
-  nofNonLiteralsInTextIndex_ = nofContexts - nofLiterals;
-  configurationJson_["num-non-literals-text-index"] =
-      nofNonLiteralsInTextIndex_;
+  configurationJson_["last-text-record-index-of-non-literals"] =
+      lastTextRecordIndexOfNonLiterals_;
   writeConfiguration();
 
   LOG(TRACE) << "END IndexImpl::passContextFileIntoVector" << std::endl;
 }
 
+// _____________________________________________________________________________
+template <typename T>
+void IndexImpl::wordsFromDocsFileEntitiesFromWordsFile(
+    const string& wordsFile, const string& docsFile,
+    const LocaleManager& localeManager, T processLine) const {
+  // Initialize DocsFileParser and WordsFileParser and the respective
+  // iterators to parse in parallel
+  auto docsFileParser = DocsFileParser{docsFile, localeManager};
+  auto docsFileIterator = docsFileParser.begin();
+  auto wordsFileParser = WordsFileParser{wordsFile, localeManager};
+  auto wordsFileIterator = wordsFileParser.begin();
+  // Get the respective lines
+  WordsFileLine currentWordsFileLine = *wordsFileIterator;
+  DocsFileLine currentDocsFileLine = *docsFileIterator;
+
+  // Iterate over docsfile and wordsfile in parallel. The wordsfile IDs are
+  // in smaller increments than the docsfile IDs in general and the
+  // wordsfile ID belongs to the next largest or equal docsfile ID. Because
+  // of this the wordsfile is advanced until the wordsfile ID is larger than
+  // the docsfile ID. During this the entities from the wordsfile are added
+  // to the respective context in this case the respective document. Once
+  // the wordsfile ID is larger than the docsfile ID the docsfile line is
+  // processed and so on.
+  while (docsFileIterator != docsFileParser.end()) {
+    currentDocsFileLine = *docsFileIterator;
```

Comment:
Can't this outer loop just be
`for (auto& currentDocsFileLine : docsFileParser)` 
(and then of course delete the begin() calls above)
Can even define the docsFileParser inline in the loop body.